### ISSUE TITLE
Effect runs when there was not change at the end

### ISSUE THREAD

**Issue by Harpush** [angular/angular#56071]
### Which @angular/* package(s) are the source of the bug?

core

### Is this a regression?

No

### Description

If I have a signal and an effect:
```ts
name = signal<string | undefined>('Angular');

effect(() => console.log(this.name()));
```
And later I do: `this.name.set('Angular');` the effect won't run as expected.

But doing (in sync):
```ts
this.name.set(undefined);
this.name.set('Angular');
```
Will trigger the effect even tough nothing actually changed.
I would expect it not to run...

### Please provide a link to a minimal reproduction of the bug

https://stackblitz.com/edit/stackblitz-starters-h9xfum?file=src%2Fmain.ts

### Please provide the exception or error you saw

```true
The effect ran and console log happened
```

### Please provide the environment you discovered this bug in (run `ng version`)

```true
Angular 17 windows 10
```

**Comment by JeanMeche** [comment]
Effects have no sense of value history. 
When a signal is updated, it marks its consummers as dirty. When an effect that consumes a signal is markAsDirty, it will schedule itself. 
This why you are observing that the effect re-run even if the signal value appears to be the same. 

If you need to tree each value independently, signal are the wrong pattern and you should look at using observables.

**Comment by Harpush** [comment]
> Effects have no sense of value history. When a signal is updated, it marks its consummers as dirty. When an effect that consumes a signal is markAsDirty, it will schedule itself. This why you are observing that the effect re-run even if the signal value appears to be the same.
> 
> If you need to tree each value independently, signal are the wrong pattern and you should look at using observables.

I wonder when is it expected for the effect to run if the signal didn't change? I mean the logic will probably execute the same as nothing actually changed.
A bit weird this is wanted...

**Comment by JeanMeche** [comment]
Effects schedule themeselves via a microtask (+ some extract behaviors in the context of components).  

Your signal did change twice:
1. `'angular` => `undefined`
2. `undefined` => `'angular'`
Both those value changes mark the consumers as dirty, because the value did actually change (`'angular'` => `'angular'` wouldn't have mark them as dirty). 

Signals are about being notified eventually when a state is dirty, not individual value updates.

**Comment by Harpush** [comment]
> Effects schedule themeselves via a microtask (+ some extract behaviors in the context of components).
> 
> Your signal did change twice:
> 
> 1. `'angular` => `undefined`
> 2. `undefined` => `'angular'`
>    Both those value changes mark the consumers as dirty, because the value did actually change (`'angular'` => `'angular'` wouldn't have mark them as dirty).
> 
> Signals are about being notified eventually when a state is dirty, not individual value updates.

That's unfortunate... Like computed won't compute if there was no change at the end I thought effect should too... I still can't see a use case for it to be honest

**Comment by JeanMeche** [comment]
Computed also re-run with the same usecase. 

https://stackblitz.com/edit/angular-next?file=src%2Fmain.ts

**Comment by Harpush** [comment]
At least it is consistent although in my opinion consistently bad. 
It means for computed as state if two updates happen which result in no change the computed will still run unnecessarily and might trigger further computed runs which will be more expensive. And you can't hail out of it.
For effect you can't rely on it as a real change notifier and must take all deps as observables with distinctUntilChanged... For example emit output on computed change must be through observable and other cases too.

To sum it - this seems unexpected

**Comment by roma2341** [comment]
Yes behavior is inconsistent, effect triggers even if my signal not changed... i checked i don't change signal, but effect is still triggered

**Comment by JeanMeche** [comment]
We'll close this as work as intended. 

Angular signals don't keep a reference to the previous value, thus can't check against them if a dirty signal has indeed changed or not.
